server/sonar-server/src/main/java/org/sonar/server/computation/ws/QualityGateStatusAction.java

```diff
@@ -0,0 +1,167 @@
+      SnapshotDto snapshotDto = getSnapshot(dbSession, ceActivityDto.getSnapshotId(), taskUuid);
+      MeasureDto measureDto = getQualityGateDetailsMeasure(dbSession, snapshotDto, taskUuid);
+
+      return buildResponse(measureDto.getData(), taskUuid);
+    } finally {
+      dbClient.closeSession(dbSession);
+    }
+  }
+
+  private CeActivityDto getCeActivity(DbSession dbSession, String taskUuid) {
+    return checkFoundWithOptional(dbClient.ceActivityDao().selectByUuid(dbSession, taskUuid),
+      "Task with id '%s' is not found", taskUuid);
+  }
+
+  private MeasureDto getQualityGateDetailsMeasure(DbSession dbSession, SnapshotDto snapshotDto, String taskUuid) {
+    List<MeasureDto> measures = dbClient.measureDao().selectBySnapshotIdAndMetricKeys(snapshotDto.getId(),
+      Collections.singleton(CoreMetrics.QUALITY_GATE_DETAILS_KEY), dbSession);
+    if (measures.isEmpty()) {
+      throw new NotFoundException(format("Quality gate details of task with id '%s' are not found", taskUuid));
+    }
+
+    return measures.get(0);
+  }
+
+  private SnapshotDto getSnapshot(DbSession dbSession, @Nullable Long snapshotId, String taskId) {
+    SnapshotDto snapshotDto = (snapshotId == null)
+      ? null
+      : dbClient.snapshotDao().selectById(dbSession, snapshotId);
+
+    return checkFound(snapshotDto, "Snapshot for task id '%s' is not found", taskId);
+  }
+
+  private static QualityGateStatusWsResponse buildResponse(String measureData, String taskUuid) {
+    try {
+      QualityGateDetails.Builder responseBuilder = QualityGateDetails.newBuilder();
+      JsonParser parser = new JsonParser();
+      JsonObject json = parser.parse(measureData).getAsJsonObject();
+
+      responseBuilder.setLevel(json.get("level").getAsString());
+      JsonArray conditions = json.getAsJsonArray("conditions");
+      QualityGateStatusWsResponse.Condition.Builder conditionBuilder = QualityGateStatusWsResponse.Condition.newBuilder();
+      Gson gson = GsonHelper.create();
+      for (JsonElement condition : conditions) {
+        responseBuilder.addConditions(conditionBuilder
+          .clear()
+          .putAllCondition(gson.<Map<String, String>>fromJson(condition, new MapJsonType().getType())));
+      }
+      return QualityGateStatusWsResponse.newBuilder().setQualityGateDetails(responseBuilder).build();
+    } catch (Exception e) {
+      throw new IllegalStateException(format("Quality gate details of task with id '%s' cannot be retrieved", taskUuid), e);
+    }
+  }
+
+  private static QualityGateStatusWsRequest toQualityGateStatusWsRequest(Request request) {
+    return new QualityGateStatusWsRequest()
+      .setTaskId(request.mandatoryParam("taskId"));
+  }
+
+  private void checkScanOrAdminPermission() {
+    if (!userSession.hasGlobalPermission(GlobalPermissions.SCAN_EXECUTION)
+      && !userSession.hasGlobalPermission(GlobalPermissions.SYSTEM_ADMIN)) {
+      userSession.checkGlobalPermission(GlobalPermissions.SYSTEM_ADMIN);
```

Comment:
throw a new ForbiddenException with the generic message Insufficent privileges. I don't want to duplicate the message, that's my only concern